preproc.c

```diff
@@ -263,7 +264,29 @@ static int emit_error_or_warning(struct tokenizer *t, int is_error) {
 static FILE *freopen_r(FILE *f, char **buf, size_t *size) {
 	fflush(f);
 	fclose(f);
-	return fmemopen(*buf, *size, "r");
```

Comment:
wouldn't it be easier to just do here `*size == 0 ? 1 : *size` ?

Reply:
Does the `*buf` pointer point to a single null byte in this case? I guess it must?

The `realloc` may not be necessary; I wasn’t sure so just tried to be safe.

What you propose might work as long as you change the mode to “w+”, but in that case it must be OK for a null byte to be written into the existing buffer.

Anything that amounts to `fmemopen(buf, 1, “r”)` will not work, because even returning a single null byte from the stream will throw off the macro location accounting, as the macro will expand to one token rather than zero. Perhaps that could be addressed with changes elsewhere, but I didn’t want to get in that deep.

Changing the mode to write-update was the only way I could find to ensure the first read returns EOF, keeping the existing code in the join function happy.